tests/integration/utils.py

```diff
@@ -162,86 +156,91 @@ def __get_reference_type(object_path, root_object_paths=None):
+        __check_attributes_schema(actual_attributes, expected_attributes)
 
     # Helper function to check through all attributes
-    def __check_attributes_schema(actual_attributes, expected_attributes,
-                                  null_value_allowed=False):
+    def __check_attributes_schema(actual_attributes, expected_attributes):
         for field, actual_value in actual_attributes.items():
             expected_attribute = expected_attributes[field]
             expected_type = __get_attribute_type(expected_attribute)
-            if not null_value_allowed or (actual_value and expected_type):
-                self.assertIsInstance(actual_value, expected_type)
+            if (actual_value and expected_type) or (
+              'nullable' in expected_attribute and
+               expected_attribute['nullable'] is False):
+                    self.assertIsInstance(actual_value, expected_type)
 
     status_code = response.status_code
     content = get_json_content(self, response)
 
-    # TODO: Add self-link testing
     # Basic tests for successful/error response
     try:
         if status_code == 200:
             resource_data = content['data']
             if isinstance(resource_data, list):
                 for resource in resource_data:
-                    __check_resource_schema(resource, null_value_allowed)
+                    __check_resource_schema(resource)
             else:
-                __check_resource_schema(resource_data, null_value_allowed)
+                __check_resource_schema(resource_data)
         elif status_code >= 400:
             errors_data = content['errors']
             self.assertIsInstance(errors_data, list)
             for error in errors_data:
-                __check_error_schema(error, null_value_allowed)
+                __check_error_schema(error)
     except KeyError as error:
         self.fail(error)
 
 
 # Check url for correct base and endpoint, parameters
 def check_url(self, actual_url, endpoint, query_params=None):
+    if query_params is None:
+        query_params = {}
+
+    base_url = self.base_url
+    if self.local_test:
+        base_url = re.sub('(:\d{4}/api)', '', self.base_url)
+
     actual_url_obj = urllib.parse.urlparse(actual_url)
-    base_url_obj = urllib.parse.urlparse(self.base_url)
-    self.assertEqual(actual_url_obj.scheme, base_url_obj.scheme)
-    self.assertEqual(actual_url_obj.netloc, base_url_obj.netloc)
-    self.assertEqual(actual_url_obj.path,
-                     f'{base_url_obj.path}{endpoint}')
-    if query_params:
-        self.assertEqual(dict(urllib.parse.parse_qsl(actual_url_obj.query)),
-                         query_params)
+    base_url_obj = urllib.parse.urlparse(base_url)
+    for actual_attribute, base_attribute, attribute_type in [
+      [actual_url_obj.scheme, base_url_obj.scheme, 'scheme'],
+      [actual_url_obj.netloc, base_url_obj.netloc, 'netloc'],
+      [actual_url_obj.path, f'{base_url_obj.path}{endpoint}', 'path']]:
```

Comment:
This is a giant array, I'd recommend to store it in another variable like [here](https://github.com/osu-mist/express-api-skeleton/pull/17/files/7fe371102dd2de9542414f9ea2067dd4ee798884..29aa5de348449f7ea951967a2b630700b51fc967#diff-7a638f68cec2ac34e29cd8f1cc2cfd24R49). That way, it will be easier to read and extend if needed in the future.